pennylane/hf/hartree_fock.py

```diff
@@ -0,0 +1,195 @@
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for performing the self-consistent-field iterations.
+"""
+
+import autograd.numpy as anp
+from pennylane.hf.matrices import (
+    generate_core_matrix,
+    generate_overlap_matrix,
+    generate_repulsion_tensor,
+    molecular_density_matrix,
+)
+
+
+def generate_scf(mol, n_steps=50, tol=1e-8):
+    r"""Return a function that performs the self-consistent-field iterations.
+
+    Args:
+        mol (Molecule): the molecule object
+        n_steps (int): the number of iterations
+        tol (float): convergence tolerance
+
+    Returns:
+        function: function that performs the self-consistent-field iterations
+
+    **Example**
+
+    >>> symbols  = ['H', 'H']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]], requires_grad = False)
+    >>> alpha = np.array([[3.42525091, 0.62391373, 0.1688554],
+    >>>                   [3.42525091, 0.62391373, 0.1688554]], requires_grad=True),
+    >>> mol = Molecule(symbols, geometry, alpha=alpha)
+    >>> args = [alpha]
+    >>> v_fock, coeffs, fock_matrix, h_core, repulsion_tensor = generate_hartree_fock(mol)(*args)
+    >>> v_fock
+    array([-0.67578019,  0.94181155])
+    """
+
+    def scf(*args):
+        r"""Perform the self-consistent-field iterations.
+
+        Args:
+            args (array[array[float]]): initial values of the differentiable parameters
+
+        Returns:
+            tuple(array[float]): eigenvalues of the Fock matrix, molecular orbital coefficients,
+            Fock matrix, core matrix
+        """
+        basis_functions = mol.basis_set
+        charges = mol.nuclear_charges
+        r = mol.coordinates
+        n_electron = mol.n_electrons
+
+        if r.requires_grad:
+            repulsion_tensor = generate_repulsion_tensor(basis_functions)(*args[1:])
+            s = generate_overlap_matrix(basis_functions)(*args[1:])
+        else:
+            repulsion_tensor = generate_repulsion_tensor(basis_functions)(*args)
+            s = generate_overlap_matrix(basis_functions)(*args)
+        h_core = generate_core_matrix(basis_functions, charges, r)(*args)
+
+        w, v = anp.linalg.eigh(s)
+        x = v @ anp.diag(anp.array([1 / anp.sqrt(i) for i in w])) @ v.T
+
+        v_fock, w_fock = anp.linalg.eigh(x.T @ h_core @ x)
+        coeffs = x @ w_fock
+
+        p = molecular_density_matrix(n_electron, coeffs)
+
+        for _ in range(n_steps):
+
+            j = anp.einsum("pqrs,rs->pq", repulsion_tensor, p)
+            k = anp.einsum("psqr,rs->pq", repulsion_tensor, p)
+
+            fock_matrix = h_core + 2 * j - k
```

Comment:
I think we should probably have a separate function that computes the Fock matrix like this

Reply:
Yes,  makes sense.